sparse/_coo/numba_extension.py

```diff
@@ -0,0 +1,165 @@
+
+    @property
+    def coords_type(self):
+        return numba.from_dtype(self.coords_dtype)[:, :]
+
+    @property
+    def shape_type(self):
+        return types.UniTuple(types.int64, self.ndim)
+
+    @property
+    def fill_value_type(self):
+        return numba.from_dtype(self.data_dtype)
+
+
+@typeof_impl.register(COO)
+def _typeof_COO(val: COO, c) -> COOType:
+    return COOType(
+        data_dtype=val.data.dtype, coords_dtype=val.coords.dtype, ndim=val.ndim
+    )
+
+
+@register_model(COOType)
+class COOModel(models.StructModel):
+    def __init__(self, dmm, fe_type):
+        members = [
+            ("data", fe_type.data_type),
+            ("coords", fe_type.coords_type),
+            ("shape", fe_type.shape_type),
+            ("fill_value", fe_type.fill_value_type),
+        ]
+        models.StructModel.__init__(self, dmm, fe_type, members)
+
+
+@type_callable(COO)
+def type_COO(context):
+    # TODO: accept a fill_value kwarg
+    def typer(coords, data, shape):
+        return COOType(
+            coords_dtype=numba.numpy_support.as_dtype(coords.dtype),
+            data_dtype=numba.numpy_support.as_dtype(data.dtype),
+            ndim=len(shape),
+        )
+
+    return typer
+
+
+@lower_builtin(COO, types.Any, types.Any, types.Any)
+def impl_COO(context, builder, sig, args):
+    typ = sig.return_type
+    coords, data, shape = args
+    coo = cgutils.create_struct_proxy(typ)(context, builder)
+    coo.coords = coords
+    coo.data = data
+    coo.shape = shape
+    coo.fill_value = context.get_constant_generic(
+        builder, typ.fill_value_type, _zero_of_dtype(typ.data_dtype)
+    )
+    return impl_ret_borrowed(context, builder, sig.return_type, coo._getvalue())
+
+
+@lower_constant(COOType)
+def lower_constant_COO(context, builder, typ, pyval):
+    coords = context.get_constant_generic(builder, typ.coords_type, pyval.coords)
+    data = context.get_constant_generic(builder, typ.data_type, pyval.data)
+    shape = context.get_constant_generic(builder, typ.shape_type, pyval.shape)
+    fill_value = context.get_constant_generic(
+        builder, typ.fill_value_type, pyval.fill_value
+    )
+    return impl_ret_borrowed(
+        context,
+        builder,
+        typ,
+        cgutils.pack_struct(builder, (data, coords, shape, fill_value)),
+    )
+
+
+@unbox(COOType)
+def unbox_COO(typ: COOType, obj: COO, c) -> NativeValue:
+    data = c.pyapi.object_getattr_string(obj, "data")
+    coords = c.pyapi.object_getattr_string(obj, "coords")
+    shape = c.pyapi.object_getattr_string(obj, "shape")
+    fill_value = c.pyapi.object_getattr_string(obj, "fill_value")
+    coo = cgutils.create_struct_proxy(typ)(c.context, c.builder)
+    coo.coords = c.unbox(typ.coords_type, coords).value
+    coo.data = c.unbox(typ.data_type, data).value
+    coo.shape = c.unbox(typ.shape_type, shape).value
+    coo.fill_value = c.unbox(typ.fill_value_type, fill_value).value
+    c.pyapi.decref(data)
```

Comment:
Is there an `xdecref`?

Reply:
Probably. but `data` will be not null so it's not needed here.

Reply:
It could be if there's an error... Or does Numba handle that?

Reply:
I'm not sure, hence https://github.com/pydata/sparse/pull/307#discussion_r368073005